### DISCUSSION TITLE
uc got detected by cloudflare on ubuntu 22.04

### DISCUSSION THREAD

**Discussion by STARKILLER-1** [seleniumbase/SeleniumBase#2522] [Q&A] [answered]
On windows working perfectly, but when we deploy it on ububtu it gets caught by cloudflare, we tried different methods from docks but failed
```
def parse_ozon_rating(articles):
    results = []
    prod = True  # Set True in production
    if prod:
        display = Display(visible=0, size=(1440, 1880))
        display.start()
        driver = Driver(uc=True, headless2=True)
    else:
        driver = Driver(uc=True)

    for index, article in enumerate(articles, start=1):
        print(f'Now article{index}/{len(articles)}')
        print('Начинаем цикл')
        attempts = 0
        while attempts < 2:
            try:
                driver.uc_open_with_tab(f"https://www.ozon.ru/product/{article}/reviews")  # driver.uc_open_with_reconnect  // driver.uc_open_with_tab
                try:
                    driver.find_element(By.CSS_SELECTOR, 'div[data-widget="error"]')
                    print('page with error, pass')
                    results.append({'article': article, 'rating': '0', 'number': '0'})
                    break
                except:
                    pass
                
                rating_span = WebDriverWait(driver, 15).until(
                    EC.presence_of_element_located((By.XPATH, "//span[contains(text(), ' / 5')]"))
                )
                rating_text = rating_span.text
                rating_value = float(rating_text.split(" / ")[0])
                print('Got ratingг')
                
                try:
                    number_div = WebDriverWait(driver, 15).until(
                        EC.presence_of_element_located((By.XPATH, '//*[@id="comments"]/div[1]')))
                    number_text = number_div.text.strip()
                    number = int(number_text)
                    print('Got reviews amount')
                except:
                    print('Error')
                    number = "Error"
                
                results.append({'article': article, 'rating': rating_value, 'number': number})
                print(results[-1])
                break 

            except Exception as e:
                print(f"Error processing article {article}: {e}")
                results.append({'article': article, 'rating': '0', 'number': '0'})
                print(results[-1])
            finally:
                attempts += 1
    
    if prod:
        display.stop()
    driver.quit()
    return results

**Comment by mdmintz** [comment, marked as answer]
Likely a change to Chrome's newer headless mode on Linux, which makes it different enough from regular Chrome to be detected. Since you already have the xvfb virtual display activated, try `headed=True`, which overrides the default headless mode on Linux. This was previously mentioned in https://github.com/seleniumbase/SeleniumBase/issues/2477#issuecomment-1934987748

**Reply by mdmintz** [reply]
@STARKILLER-1 Also try again with the latest version of SeleniumBase. https://github.com/seleniumbase/SeleniumBase/issues/2523 was found and fixed. That may have been the root issue. Let me know if that fixes headless UC Mode.

**Reply by STARKILLER-1** [reply]
@mdmintz We have used 4.20.9 seleniumbase version with headed=True and it worked! But site's own detection system caught us. I think we will manage this by clicking "Обновить"- means refresh
![image](https://github.com/seleniumbase/SeleniumBase/assets/102546596/2cee9fc2-288d-4107-a04b-af7ebc981584)
translation:
Access is restricted
Incident: fab_chlg_20240220173155_01HQ3S5YVR9ZBYF30V3MXCEA42

To solve the problem, try to do this:

Wait a bit and click on the "Update" button
Disable VPN if it is in use
Update the browser or mobile app version
Connect to another WI-FI or mobile network
Restart your home router if you are using home WI-FI

**Reply by STARKILLER-1** [reply]
Also we will update sb and retry

**Reply by STARKILLER-1** [reply]
@mdmintz This button is a trap of cloudflare when we click it, we are getting detected
``try:
                    button = WebDriverWait(driver, 3).until(
                        EC.presence_of_element_located((By.XPATH, "//button[contains(text(), 'Обновить')]"))
                    )
                    button.click()
                    time.sleep(2)
                except Exception as e:
                    print(e)``

**Reply by mdmintz** [reply]
Upgrade to the latest `seleniumbase`, use `driver.uc_open_with_reconnect(url, reconnect_time)` with a longer `reconnect_time` as needed, and use `driver.uc_click(selector)` instead of regular clicks to avoid detection.

**Reply by STARKILLER-1** [reply]
*Upgraded to the latest seleniumbase*
'WebDriver' object has no attribute 'uc_open_with_reconnect'
```
def parce_ozon_rating(articles):
    results = []
    prod = True 
    if prod:
        display = Display(visible=0, size=(1440, 1880))
        display.start()
        driver = Driver(uc=True, headed=True)
    else:
        driver = Driver(uc=True)
    for index, article in enumerate(articles, start=1):
        attempts = 0
        while attempts < 2:
            try:
                driver.uc_open_with_reconnect(f"https://www.ozon.ru/product/{article}/reviews", 5)  # driver.uc_open_with_reconnect  // driver.uc_open_with_tab
                try:
                    button = WebDriverWait(driver, 3).until(
                        EC.presence_of_element_located((By.XPATH, "//button[contains(text(), 'Обновить')]"))
                    )
                    send_telegram_notification('Кнопка обновления найдена')
                    button.click()
                    #driver.uc_click("//button[contains(text(), 'Обновить')]") #also 'WebDriver' object has no attribute uc_click
                    time.sleep(2)
                except Exception as e:
```

**Reply by mdmintz** [reply]
This is working for me on the latest version:

```python
from seleniumbase import Driver

driver = Driver(uc=True, headed=True)
try:
    driver.uc_open_with_reconnect("https://www.ozon.ru/product/", 5)
finally:
    driver.quit()
```

**Reply by mdmintz** [reply]
The `driver.uc_open_with_reconnect()` method gets added when using `uc=True`.

**Reply by STARKILLER-1** [reply]
Thank you, that was one-time bug
How can I uc_click on element using xpath, not css selector(because there is a security which makes all attributes different on different pages)
```
<button class="rb" id="reload-button">Обновить</button>
        <a href="https://www.ozon.ru/complaint/support/?incident_id=fab_chlg_20240220173155_01HQ3S5YVR9ZBYF30V3MXCEA42&amp;platform=web&amp;back_url=/">
            <button class="sb" tabindex="-1">Служба поддержки</button>
        </a>
    </div>
```
Xpath= "//button[contains(text(), 'Обновить')]"
class and id are changing every re-open of the page

**Reply by mdmintz** [reply]
Set `by="xpath"` (`by="css selector"` is the default value)
Most methods let you set/change that.

**Reply by STARKILLER-1** [reply]
All worked, I am re-opening driver on each page and it's working for me, When i tried to open multiple pages one after another in a single driver, cloudflare catches my script

**Reply by STARKILLER-1** [reply]
Thanks to your work, I can train juniors to work with the most advanced framework, which has everything you need

**Comment by mdmintz** [comment]
Lots of improvements to UC Mode recently. Upgrade to `4.28.1` (or newer) to bypass CF on Linux. More info: https://github.com/seleniumbase/SeleniumBase/issues/2865
